tests/test_peak_core_and_tile.py

```diff
@@ -0,0 +1,197 @@
+    * drives input values onto data0 and data1 ports
+    * checks alu_res output
+    """
+    def core_input_driver(tester, value):
+        config_data, a, b, = value
+        for addr, data in config_data:
+            tester.configure(addr, data)
+        tester.poke(tester._circuit.data0, a)
+        tester.poke(tester._circuit.data1, b)
+
+    def core_output_monitor(tester, value):
+        tester.expect(tester._circuit.alu_res, value)
+
+    core = PeakCore(PE_fc)
+    core.name = lambda: "PECore"
+    circuit = core.circuit()
+
+    inputs, outputs = sequences
+    tester = BasicSequenceTester(circuit, [
+        (core_input_driver, InputSequence(inputs)),
+        (core_output_monitor, OutputSequence(outputs))
+    ], circuit.clk, circuit.reset)
+    tester.reset()
+
+    with tempfile.TemporaryDirectory() as tempdir:
+        for filename in dw_files:
+            shutil.copy(filename, tempdir)
+
+        tester.compile_and_run("verilator",
+                               directory=tempdir,
+                               magma_opts={"coreir_libs": {"float_DW"}},
+                               flags=["-Wno-fatal"])
+
+
+def test_peak_tile_sequence(sequences, dw_files):
+    """
+    Tile level test:
+    * Generates CGRA to extract PE_tile
+    * sets tile_id input based on test app
+    * configures PE_tile based on fixed route (taken from a test app)
+    * similar input driver and output monitor behavior to core test except:
+      * inputs are driven onto the appropriate tile ports based on the fixed
+        route configuration
+      * output is similarly monitored based on the route config
+    """
+    # Use stub CGRA to get PE_tile
+    # TODO: Is there an API to just get a PE_tile?
+    io_sides = IOSide.North | IOSide.East | IOSide.South | IOSide.West
+    interconnect = create_cgra(2, 2, io_sides, num_tracks=3)
+
+    # Use test application to get a configuration
+    # TODO: We should randomize or cover all possible configurations of the
+    # CB/SBs
+    # placement, routing generated from:
+    # netlist = {
+    #     "e0": [("I0", "io2f_16"), ("p0", "data0")],
+    #     "e1": [("I1", "io2f_16"), ("p0", "data1")],
+    #     "e3": [("p0", "alu_res"), ("I2", "f2io_16")],
+    # }
+    # bus = {"e0": 16, "e1": 16, "e3": 16}
+
+    # from archipelago import pnr
+    # placement, routing = pnr(interconnect, (netlist, bus))
+    # route_config = interconnect.get_route_bitstream(routing)
+    # x, y = placement["p0"]
+
+    # placement = {'I0': (1, 0), 'I1': (2, 0), 'I2': (0, 1), 'p0': (1, 1)}
+    # routing = {'e0': [[CB_io2f_16, SB_T0_NORTH_SB_IN_B16, CB_data0]],
+    #            'e1': [[CB_io2f_16, SB_T0_NORTH_SB_IN_B16,
+    #                    SB_T0_WEST_SB_OUT_B16, RMUX_T0_WEST_B16,
+    #                    SB_T0_EAST_SB_IN_B16, CB_data1]],
+    #            'e3': [[CB_alu_res, SB_T0_WEST_SB_OUT_B16, RMUX_T0_WEST_B16, CB_f2io_16]]}
+
+    route_config = [(262401, 0), (459265, 0), (459265, 0), (327937, 2),
```

Comment:
All possible CB/SB combination is tested in canal, so I don't think you need to do it here. I'd suggest use more readable netlist form and have PnR tool to figure out the proper route. If the default CGRA interconnect parameter changes, this bitstream won't work anymore.

Reply:
Is the routing guaranteed to be the same every time I run the test? Or is there a way to forced this with a seed? When I revert to using the PnR tool to use the test, it seems I get some occasional failures.

I added
```
    assert route_config == ([(262401, 0), (459265, 0), (459265, 0),
                             (327937, 2), (459009, 786432), (459009, 0)])
```
to check and it seems that occasionally it produces a different configuration:
```
E       assert [(262401, 0),..., (459009, 0)] == [(262401, 0),..., (459009, 0)]
E         At index 1 diff: (17236481, 0) != (459265, 0)
E         Use -v to get the full diff
```

If we add the API to get the port name from PnR, then I think we can avoid this problem since the actual configuration won't matter.

We don't need to do full coverage if you're already doing it, but doing some randomized tests would be nice (this also just serves as an example for other users of the tool as to how to do randomized tests and lifting from core to tile level).  This also needs the API for port name from PnR, so I think that would be the ideal solution.

Reply:
I believe the PnR should be deterministic. This sounds like a bug. I will look into it.

Randomizing PnR can be difficult since the entire toolchain is designed to optimize delay and switching power. If there is only one optimal result, then very likely you'll get the same result.

What I did for testing is actually manually enumerate all possible routes between two IO ports. If you want to follow that route, I can provide an API that give you all possible routes and you can randomly pick one.

Reply:
I can't seem to reproduce the randomness you're experiencing. I tested with two different linux environments for 100 iterations, although I did notice inconsistent result on different machines. I have some idea of where this inconsistent behavior might come from.

Reply:
>  I can provide an API that give you all possible routes and you can randomly pick one.

that sounds good

> I can't seem to reproduce the randomness you're experiencing. I tested with two different linux environments for 100 iterations, although I did notice inconsistent result on different machines. I have some idea of where this inconsistent behavior might come from.

let me know if there's anything I can do on my machine to help debug.

I'm seeing this captured output, maybe the "unable to determine tmax" is the problem?
```
4.000000 -> 4.000000 improvement: 0.000000 total: 0.000000
using bit_width 1
Routing iteration:   0 duration: 0 ms
using bit_width 16
Routing iteration:   0 duration: 0 ms
-------------------------------------------------- Captured stderr call ---------------------------------------------------
Unable to determine tmax. Use default temperature
```

Reply:
And it also shows the "Unable to determine tmax. Use default temperature" so that's probably not the problem

Reply:
> And it also shows the "Unable to determine tmax. Use default temperature" so that's probably not the problem
This is just a warning at the last placement stage where placement are further optimized based on annealing algorithm, hence the term temperature. It tries to figure out the approximate annealing schedule first. If it fails to do so, it will just use the default temperature.

> It seems like travis/linux gets the same result as the assertion failure I showed above: https://travis-ci.com/github/StanfordAHA/garnet/builds/189703333, not sure if that's helpful
I think all linux systems should give
```
[(262401, 0), (17236481, 0), (459265, 0), (327937, 10), (459009, 786432), (459009, 0)]
```
Can you try the latest release of PnR on your machine? I think the update should make all PnR result consistent across different platforms.

Reply:
Ok, that seems to resolve the issue based on a couple runs, I'll let you know if it comes up.  I left an assertion with the expected values in because if the route does change, we need to update the test (in case the ports that are used changed).  We can remove the assertion once we add/use the API to get the tile input/output ports that are used.